### PR TITLE
Parse arguments even without 'action'.

### PR DESCRIPTION
**Pull request by IvanGoncharov**
In case if no arguments provided there was no error about missing argument.
Also arguments was validated if no action specified.

[Review on Reviewable](https://reviewable.io/reviews/tj/commander.js/405)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -602,26 +560,57 @@ Command.prototype.normalize = function(args) {
  */
 
 Command.prototype.parseArgs = function(args, unknown) {
-  var name;
+  var self = this;
 
-  if (args.length) {
```
Comment (IvanGoncharov) [inline index.js]: No check for required args, if no arguments provided.

File: index.js [hunk 2]
```diff
@@ -257,53 +257,13 @@ Command.prototype.parseExpectedArgs = function(args) {
 
 Command.prototype.action = function(fn) {
   var self = this;
-  var listener = function(args, unknown) {
```
Comment (IvanGoncharov) [inline index.js]: Check argument only if 'action' registered.

File: test/test.variadic.args.js [hunk 3]
```diff
@@ -26,7 +26,10 @@ program.args.should.have.lengthOf(3);
 program.args[0].should.eql('arg0');
 program.args[1].should.eql(['arg1', 'arg2', 'arg3']);
 
-program
+//Create fresh instance
+var program2 = new program.Command();
+
+program2
```
Comment (IvanGoncharov) [inline test/test.variadic.args.js]: same sub command was registered twice.
Previously it work because events was overridden.
But inside `commands` array it was two instances with same name.
This lead to problem with help messages and should be handled in separate commit.
IMHO, exception should be thrown then duplicated commands registered.

**Comment by IvanGoncharov** [comment]
@zhiyelee Did you have a chance to look at my changes?
It pretty simple and safe, because most of changes is code move from one function into another.
If required I can try split them into several commits.

**Comment by SomeKittens** [comment]
@IvanGoncharov it'd help if you could summarize the why (not what) of this PR

**Comment by IvanGoncharov** [comment]
Main reason is following scenario:
```javascript
#!/usr/bin/env node
var program = require('./index.js');

program
  .arguments('<ARG>')
  .parse(process.argv);
```
Before changes:
>$ ./test.js
>$

After my changes:
>$ ./test.js 
>
>  error: missing required argument `ARG'

**Comment by IvanGoncharov** [comment]
It could be workaround simply be removing [this](https://github.com/tj/commander.js/blob/master/index.js#L607) check.
But I think delaying args parsing until event handler unreasonably complicate code.
It also force register events for aliases and create complicated wrapper for listener.

**Comment by SomeKittens** [comment]
This is a pretty big change - not sure if I want to make it at this point.  @zhiyelee ?

**Comment by IvanGoncharov** [comment]
When I debug these issue, I spent quite some time just to understand where args parsing is happening.
I never would have guessed, it resides inside listener wrapper.
The simpler and clean code is, the more people would contribute.
That is my two cents.

To simplify review, I can break it into several commits.

**Comment by zhiyelee** [comment]
@SomeKittens Really big change, I will review it later this weekend.

**Comment by tvervest** [comment]
Any news on this PR? I'd love to see built-in required arguments :+1:
